Experimental/src/Ranges/Range.cs

```diff
@@ -0,0 +1,713 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using JetBrains.Annotations;
+
+namespace CodeJam.Ranges
+{
+    [Flags]
+    public enum RangeOptions
+    {
+        None = 0,
+        HasStart = 0x1,
+        HasEnd = 0x2,
+        IncludingStart = 0x4,
+        IncludingEnd = 0x8,
+        IsEmpty = 0x10,
+    }
+
+	[PublicAPI]
+    public static class Range
+    {
+        public static Range<TValue> Create<TValue>(TValue start, TValue end, bool includeStart, bool includeEnd)
+            where TValue : IComparable<TValue>
+        {
+            return new Range<TValue>(start, end,
+                RangeOptions.HasStart | RangeOptions.HasEnd |
+                (includeStart ? RangeOptions.IncludingStart : RangeOptions.None) |
+                (includeEnd ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+		public static Range<TValue> Create<TValue>(TValue start, TValue end, bool include = true)
+            where TValue : IComparable<TValue>
+        {
+            return new Range<TValue>(start, end,
+                RangeOptions.HasStart | RangeOptions.HasEnd |
+                (include ? RangeOptions.IncludingStart | RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public static Range<TValue> StartsWith<TValue>(TValue start, bool include = true)
+            where TValue : IComparable<TValue>
+        {
+            return new Range<TValue>(start, default(TValue),
+                RangeOptions.HasStart | (include ? RangeOptions.IncludingStart : RangeOptions.None));
+        }
+
+        public static Range<TValue> Simple<TValue>(TValue value)
+            where TValue : IComparable<TValue>
+        {
+            return Create(value, value, true, true);
+        }
+
+        public static Range<TValue> EndsWith<TValue>(TValue end, bool include = true)
+            where TValue : IComparable<TValue>
+        {
+            return new Range<TValue>(default(TValue), end,
+                RangeOptions.HasEnd | (include ? RangeOptions.IncludingEnd : RangeOptions.None));
+        }
+
+        public static Range<TValue> Empty<TValue>()
+            where TValue : IComparable<TValue>
+        {
+            return Range<TValue>.Empty;
+        }
+
+        public static Range<TValue> Full<TValue>()
+            where TValue : IComparable<TValue>
+        {
+            return Range<TValue>.Full;
+        }
+    }
+
+	[PublicAPI]
+	[DebuggerDisplay("{DisplayValue()}")]
+    public struct Range<TValue> : IComparable<Range<TValue>>
+        where TValue : IComparable<TValue>
+    {
+		public static readonly Range<TValue> Empty = new Range<TValue>(default(TValue), default(TValue),
+            RangeOptions.IsEmpty);
+
+        public static readonly Range<TValue> Full = new Range<TValue>(default(TValue), default(TValue),
+            RangeOptions.None);
+
+        private readonly RangeOptions _options;
+        public readonly TValue End;
```

Comment:
So, for reference types it is not really immutable. This should be stated in the docs